File: saif_erp/saif_erp/report/employee_document_expiry/employee_document_expiry.js
// Copyright (c) 2026, SGA World FZ LLC and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Employee Document Expiry"] = {
	filters: [
		{ fieldname: "company", label: __("Company"), fieldtype: "Link", options: "Company" },
		{
			fieldname: "document", label: __("Document"), fieldtype: "Select",
			options: ["", "Passport", "Residence Visa", "Emirates ID", "Labour Card", "Health Insurance"],
		},
		{
			fieldname: "band", label: __("Status"), fieldtype: "Select",
			options: ["", "Expired", "≤ 30 days", "31 – 60 days", "61 – 90 days", "> 90 days"],
		},
		{ fieldname: "only_expiring", label: __("Only expiring (≤ 90 days)"), fieldtype: "Check", default: 0 },
	],

	// colour Days Left (bold number) + Status (filled pill) by urgency
	formatter(value, row, column, data, default_formatter) {
		const colors = {
			"Expired": "#8b1a1a", "≤ 30 days": "#e0533d", "31 – 60 days": "#e8804d",
			"61 – 90 days": "#e6a817", "> 90 days": "#22a06b",
		};
		const c = (data && colors[data.band]) || "#888";
		if (data && column.fieldname === "band") {
			return `<span style="color:#fff;background:${c};padding:2px 9px;border-radius:11px;font-weight:600;white-space:nowrap">${frappe.utils.escape_html(data.band || "")}</span>`;
		}
		if (data && column.fieldname === "days_left") {
			return `<span style="color:${c};font-weight:700">${data.days_left}</span>`;
		}
		return default_formatter(value, row, column, data);
	},
};

File: saif_erp/saif_erp/report/employee_document_expiry/employee_document_expiry.py
# Copyright (c) 2026, SGA World FZ LLC and contributors
# For license information, please see license.txt
"""Employee Document Expiry — passport / visa / Emirates ID / labour card / insurance
expiry, colour-coded by urgency (expired -> red, > 90 days -> green)."""

from collections import Counter

from saif_erp.document_expiry import BANDS, get_expiring_documents


def execute(filters=None):
	filters = filters or {}
	rows = get_expiring_documents(company=filters.get("company"))

	if filters.get("document"):
		rows = [r for r in rows if r["document"] == filters.get("document")]
	if filters.get("band"):
		rows = [r for r in rows if r["band"] == filters.get("band")]
	if filters.get("only_expiring"):
		rows = [r for r in rows if r["days_left"] <= 90]

	counts = Counter(r["band"] for r in rows)
	return _columns(), rows, None, _chart(counts), _summary(counts)


def _columns():
	return [
		{"label": "Employee", "fieldname": "employee", "fieldtype": "Link", "options": "Employee", "width": 120},
		{"label": "Name", "fieldname": "employee_name", "fieldtype": "Data", "width": 180},
		{"label": "Company", "fieldname": "company", "fieldtype": "Data", "width": 190},
		{"label": "Document", "fieldname": "document", "fieldtype": "Data", "width": 130},
		{"label": "Number", "fieldname": "number", "fieldtype": "Data", "width": 140},
		{"label": "Expiry Date", "fieldname": "expiry", "fieldtype": "Date", "width": 110},
		{"label": "Days Left", "fieldname": "days_left", "fieldtype": "Int", "width": 90},
		{"label": "Status", "fieldname": "band", "fieldtype": "Data", "width": 120},
	]


def _chart(counts):
	labels = [b[0] for b in BANDS]
	return {
		"type": "bar",
		"data": {"labels": labels, "datasets": [{"name": "Documents", "values": [counts.get(b[0], 0) for b in BANDS]}]},
		"colors": ["#e0533d"],
	}


def _summary(counts):
	summary = [
		{"value": counts.get(label, 0), "label": label, "indicator": ind, "datatype": "Int"}
		for label, lo, hi, color, ind in BANDS
	]
	summary.append({"value": sum(counts.values()), "label": "Total tracked",
	                "indicator": "Grey", "datatype": "Int"})
	return summary
